Libraries/Cordio/wsf/sources/targets/baremetal/wsf_nvm.c

```diff
@@ -172,63 +195,72 @@ void WsfNvmInit(void)
-
-      if (headerCrc != header.headerCrc)
-      {
-        /* Corrupt header. */
-        /* TODO: Catastrophic failure? */
-        break;
-      }
-      else if ((header.id == id) && (header.len == len))
-      {
-        /* Valid header and matching ID - read data after header. */
-        storageAddr += sizeof(header);
-        PalFlashRead(pData, header.len, storageAddr);
-        dataCrc = CalcCrc32(WSF_NVM_CRC_INIT_VALUE, header.len, pData);
-        if (dataCrc == header.dataCrc)
-        {
-          findId = TRUE;
-        }
-        break;
-      }
-    }
+    WSF_ASSERT(!((id == WSF_NVM_RESERVED_FILECODE) || (id == WSF_NVM_UNUSED_FILECODE)));
 
-    /* Move to next stored data block and read header. */
-    storageAddr += WSF_NVM_WORD_ALIGN(header.len) + sizeof(header);
+    /* Read first header. */
     PalFlashRead(&header, sizeof(header), storageAddr);
-  } while(1);
 
-  if (compCback)
-  {
-    compCback(findId);
-  }
-  return findId;
+    do {
+        if (header.id == WSF_NVM_UNUSED_FILECODE) {
+            /* Found unused entry at end of used storage. */
+            break;
+        }
+
+        /* Iterate through stored data headers, looking for existing matching stored data header. */
+        if (header.id != WSF_NVM_RESERVED_FILECODE) {
+            /* Calculate CRC of header itself. */
+            headerCrc = CalcCrc32(WSF_NVM_CRC_INIT_VALUE, sizeof(header.id) + sizeof(header.len),
+                                  (uint8_t *)&header);
+
+            if (headerCrc != header.headerCrc) {
+                /* Corrupt header. */
+                /* TODO: Catastrophic failure? */
+                break;
+            } else if ((header.id == id) && (header.len == len)) {
+                /* Valid header and matching ID - read data after header. */
+                storageAddr += sizeof(header);
+                PalFlashRead(pData, header.len, storageAddr);
+                dataCrc = CalcCrc32(WSF_NVM_CRC_INIT_VALUE, header.len, pData);
+                if (dataCrc == header.dataCrc) {
+                    findId = TRUE;
+                }
+                break;
+            }
+        }
+
+        /* Move to next stored data block and read header. */
+        storageAddr += WSF_NVM_WORD_ALIGN(header.len) + sizeof(header);
+        PalFlashRead(&header, sizeof(header), storageAddr);
+    } while (1);
+
+    if (compCback) {
+        compCback(findId);
+    }
+    return findId;
 }
 
+/*************************************************************************************************/
+/*!
+ *  \brief  Check whether or not the amount of data has space to fit into NVM
+ *
+ *  \param  lenNeeded         Number of bytes wanting to store
+ *
+ *  \return TRUE if NVM operation is successful, FALSE otherwise.
+ */
+/*************************************************************************************************/
+static inline bool_t wsfNvmHaveEnoughSpace(uint32_t lenNeeded)
+{
+    if (lenNeeded > WsfNvmGetRemainingSpace()) {
```

Comment:
This comparison between unsigned `lenNeeded` and the signed return value of `WsfNvmGetRemainingSpace()` can be problematic in some cases. When `WsfNvmGetRemainingSpace()` is negative, this expression evaluates to `FALSE`, which causes the calling code to be unaware that there is not enough space for the write to succeed.

Changing `lenNeeded` to an `int32_t` seems to have the desired effect.

Reply:
As mentioned in the previous comment, I updated the code to return an unsigned value, removing this issue.